R/pivot.R

```diff
@@ -0,0 +1,139 @@
+#' }
+#' @author Daniel Falster - daniel.falster@unsw.edu.au
+#' @export
+trait_pivot_wider <- function(traits) {
+
+  metadata_cols <- c("unit", "replicates", "measurement_remarks", "basis_of_value")
+
+  # A check for if there are more than 1 value_type for a given taxon_name, observation_id and method
+  check_value_type <- traits %>%
+    select(dplyr::all_of(c(
+      "trait_name", "value", "dataset_id", "observation_id", "method_id", "method_context_id",
+      "repeat_measurements_id", "value_type"))) %>%
+    dplyr::group_by(
+      .data$dataset_id, .data$observation_id, .data$method_id,
+      .data$method_context_id, .data$repeat_measurements_id) %>%
+    dplyr::summarise(n_value_type = length(unique(.data$value_type))) %>%
+    arrange(.data$observation_id) %>%
+    dplyr::filter(.data$n_value_type > 1)
+
+  if (nrow(check_value_type) > 1) {
+
+    traits %>%
+      select(-dplyr::all_of(metadata_cols)) %>%
+      # Sophie - what's the point of the group_by?
+      dplyr::group_by(
+        .data$dataset_id, .data$observation_id, .data$method_id,
+        .data$method_context_id, .data$repeat_measurements_id, .data$value_type) %>%
+      tidyr::pivot_wider(names_from = "trait_name", values_from = "value") |>
+      dplyr::ungroup()
+
+  } else {
+
+    metadata_cols <- c(metadata_cols, "value_type")
+
+    traits %>%
+      select(-dplyr::all_of(metadata_cols)) %>%
+      # Sophie - what's the point of the group_by?
+      dplyr::group_by(
+        .data$dataset_id, .data$observation_id, .data$method_id, .data$method_context_id,
+        .data$repeat_measurements_id) %>%
+      tidyr::pivot_wider(names_from = "trait_name", values_from = "value") |>
+      dplyr::ungroup()
+  }
+
+}
+
+
+#' @title Pivot wide format data into a long format
+#'
+#' @description `trait_pivot_longer` "gathers" wide format data into a "tidy" format.
+#'
+#' This function converts the data into long format where observations are on different rows and the type of
+#' observation is denoted by the `trait_name` column.
+#' In other words, `trait_pivot_longer` reverts the actions of `trait_pivot_wider`.
+#' @param wide_data Output from `trait_pivot_wider` (a tibble of wide data)
+#' @return A tibble in long format
+#' @details
+#' `trait_pivot_longer` will return a tibble with fewer columns than the original traits table
+#' The excluded columns include: "unit", "replicates", "measurement_remarks", "basis_of_record",
+#' "basis_of_value" # Double check #TODO
+#'
+#' @examples
+#' \dontrun{
+#' data <- austraits$traits %>%
+#' filter(dataset_id == "Falster_2003")
+#' data # Long format
+#' traits_wide <- trait_pivot_wider(data)
+#' traits_wide # Wide format
+#'
+#' values_long <- trait_pivot_longer(traits_wide)
+#' }
+#' @author Daniel Falster - daniel.falster@unsw.edu.au
+#' @author Fonti Kar - fonti.kar@unsw.edu.au
+#' @export
+trait_pivot_longer <- function(wide_data) {
+
+  wide_data %>%
+    tidyr::pivot_longer(
+      cols = 20:ncol(.),
```

Comment:
I think in the documention we should describe where the `20` comes from, because eventually we'll forget that this is the number of "non-trait-value" columns in the wide dataset.

Reply:
Good idea, added a comment